keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1108 @@
+
+For all objects other than Applications, upon object deletion, the Application Controller shall do the following. 
+
+1. [Retrieve the Application for the object](#retrieving-applications).
+1. Update the Application's `Status` to indicate that component is no longer installed.
+
+For all Application objects, upon deletion, the Application Controller will do the following.
+
+1. List all Applications in the same namespace as the application under consideration.
+1. For all Applications that contain a reference to the Application under consideration in their `Depepndecies`, the 
+Application Controller will update the Application's `Statuts` to indicate that the Application under consideration is 
+no longer installed.
+
+For all objects other than Applications, when an object is updated, the Application Controller will do the following.
+
+1. [Retrieve the Application for the object](#retrieving-applications).
+1. If the Application's `Selector` does not match the object's labels, the Application Controller will update the 
+Applications `Status` to indicate that component is no longer installed and remove any `OwnerReference` from the 
+Application to the object.
+
+#### Health Status Reporting.
+The Application Controller will do the following to update the `Stauts.Ready` ConditionStatus of an Application object.
+
+1. If the Application Controller observes an Application object with a nil `Spec.HealthCheckDeployment`, it shall set 
+the `Status.Ready` field of the Application object to `UNKNOWN`.
+1. If the Application Controller observes a Deployment created from an ApplicationHealthCheck it shall 
+[retrieve the Application for the object](#retrieving-applications).
+   1. If the Deployment is complete and healthy, the Application Controller shall set the `Status.Ready` field of the 
+   Application's `.Status` to `True`.
+   object to `UKNOWN`.
+   1. If the Pod has a Ready Condition, the Application Controller shall set the `Status.Ready` field of the Application 
+   object to the Status of the Pod's Ready Condition.
+
+#### Generation Observation
+Each time the Application Controller observes the creation or mutation of an Application object, it will set the 
+`Status.ObservedGeneration` to the `Generation` of the Application. As monotonicity of the object's `Generation` is 
+ensured by the API Server, the `Satatus.ObservedGeneration` can be used to determine if the Application Controller 
+has observed a mutation to an Application object.
+
+#### Adopting Existing Components
+The design of [Garbage Collection](#garabage-collection) allows existing objects to be aggregated into an Application. 
+In order to achieve this users should do the following.
+
+1. Create an Application object that contains the objects that will be adopted in the Application's `Spec.Components`.
+1. Annotate the objects with a `kubernetes.io/application` label that references the Application.
```

Comment:
Is this recursive?  Meaning, if I mention a Deployment as a Component, and the Deployment makes and RS that makes a pod, does the pod have the label too?  What about a SparkJob, where there is no visible pod template in the parent resource?

Also, "Annotate ... with ... label" makes it unclear if you are talking about a metadata.labels or a metadata.annotations.